New How to run AM3 docs (#1238)
ACCESS-Hive/access-hive.github.io: docs/includes/rose_cylc.md

[gadi]: https://opus.nci.org.au/display/Help/0.+Welcome+to+Gadi#id-0.WelcometoGadi-Overview
[PBS job]: https://opus.nci.org.au/display/Help/4.+PBS+Jobs
# Run models using Rose/Cylc

!!! warning

>     ACCESS models configurations that run with _Rose/Cylc_ currently use _Cylc7_, with intentions to upgrade to _Cylc8_. The upgrade is expected to change some aspects of the workflow described on this page. Updated information about the _Cylc8_ workflow will be provided once a model configuration using this version becomes available.

Reviewer: ```suggestion
    ACCESS model configurations that run with _Rose/Cylc_ currently use _Cylc7_, with plans to upgrade to _Cylc8_. This upgrade is expected to change some aspects of the workflow described on this page. Information on the _Cylc8_ workflow will be provided once a model configuration using this version is available.
```

## About
The _Rose/Cylc_ workflow management tool consists of two components:

> * The [_Cylc_](https://niwa.co.nz/environmental-information/cylc-suite-engine) (pronounced ‘silk’) task engine, developed by the New Zealand National Institute of Water and Atmospheric Research (NIWA). _Cylc_ is a workflow manager that automatically executes tasks according to the model's configuration. It also monitors all tasks, reporting any errors that may occur.

Reviewer: I would add the acronym for NIWA in `abbreviations.md` file rather than defining it here:
 `ACCESS-Hive-Docs/
└── abbreviations.md`
```suggestion
* The [Cylc](https://niwa.co.nz/environmental-information/cylc-suite-engine) (pronounced ‘silk’) task engine, developed by NIWA, is a workflow manager that automatically executes tasks according to a model configuration and monitors them for errors.
```

* The [_Rose_](https://www.metoffice.gov.uk/research/approach/modelling-systems/rose) framework developed by the UKMO configures tasks for the _Cylc_ engine. _Rose_ is a toolkit that can be used to view, edit and run some of the [ACCESS models](/models/access_models).

> A set of tasks configured by _Rose_ to run with the _Cylc7_ engine is called a _suite_. Every _suite_ has a unique identifier called `suite-ID` in the form `u-LLNNN`, where `L` is a letter and `N` is a number (e.g., u-ab123).

Reviewer: Did you want to be specific here with `Cylc7` or more general to include `Cylc8`?

    Once you are connected, skip directly to [Set up a persistent session](#set-up-a-persistent-session).

??? info "Connect via ARE VDI Desktop"

    If you are not familiar with ARE, check out the [Getting Started on ARE](/getting_started/are) page.

    !!! tip
        The ARE VDI session does not run model tasks directly; it only runs _Rose/Cylc_. The model tasks are dispatched by _Cylc_ to the compute nodes. This means the ARE VDI session requires minimal CPU and memory resources.

    Go to the [ARE VDI](https://are.nci.org.au/pun/sys/dashboard/batch_connect/sys/desktop_vnc/ncigadi/session_contexts/new) page and launch a session with the following entries:

    - **Walltime (hours)** &rarr; `2`<br>
            Amount of hours the ARE VDI session will remain active for. This is only setup time, and does not reflect how long the actual configuration will take to run.
            
        !!! tip 

>             Some model configurations might require a longer setup time. The _Walltime_ included here should be sufficient for most model configurations, but if your ARE session terminates before the model setup is complete, you can start a new ARE VDI session.

Reviewer: ```suggestion
            Some model configurations may require additional setup time. The _Walltime_ specified here should be sufficient for most configurations, but if your ARE session terminates before setup is complete, you can start a new ARE VDI session and continue.
```

The ACCESS models which use _Cylc_ require a connection to the MOSRS mirror on _Gadi_. To connect to this mirror, you must first authenticate your MOSRS credentials with:

```
mosrs-auth
```

This will request the username and password you received when you created your MOSRS account.

<terminal-window>
    <terminal-line data="input">mosrs-auth</terminal-line>
    <terminal-line lineDelay=500><span style="color: #559cd5;">INFO</span>: You need to enter your MOSRS credentials here so that GPG can cache your password.</terminal-line>
    <terminal-line>Please enter the MOSRS password for &lt;MOSRS-username&gt;:</terminal-line>
    <terminal-line lineDelay=1500>Checking your credentials using Subversion. Please wait.</terminal-line>
    <terminal-line lineDelay=500><span style="color: #559cd5;">INFO</span>: Successfully accessed Subversion with your credentials.</terminal-line>
    <terminal-line lineDelay=100><span style="color: #559cd5;">INFO</span>: Checking your credentials using rosie. Please wait.</terminal-line>
    <terminal-line lineDelay=500><span style="color: #559cd5;">INFO</span>: Successfully accessed rosie with your credentials.</terminal-line>
</terminal-window>

> After the first authentication, you will need to run `mosrs-auth` every 24 hours and for every new connection to _Gadi_ (e.g., new terminal) to verify your password against the saved credentials.

Reviewer: ```suggestion
After the initial authentication, run `mosrs-auth` every 24 hours and after each new connection to _Gadi_ (e.g., a new terminal) to verify your password against the saved credentials.
```

## Get the model configuration

Depending on the specific model, its configuration will be hosted either on _GitHub_ or MOSRS. The [Run a Model](/models/run_a_model/) documentation for the respective model will specify where the configuration is stored.<br>

### Model configurations stored on _GitHub_

<!--start:get-github-config-->
To get a local copy of the configuration of your choice, clone the _GitHub_ branch with:

```
git -C ~/roses clone {{github_configs}} -b <branch> <experiment_name>
```

where:

- `<branch>` is the name of the branch of the configuration you want to base your work on.

> - `<experiment_name>` is the name of your local copy of the configuration, i.e. your _configuration directory_. It will also be used by _Cylc_ for your _control_ directory path.

Reviewer: ```suggestion
- `<experiment_name>` is the name of your local copy of the configuration, i.e., your _configuration directory_. _Cylc_ also uses it for the _control_ directory path.
```

!!! warning
    Configuration repositories store each configuration in their own branch, while the `main` branch only keeps minimal setup.<br>
    For this reason, when forking the repository, ensure that all branches are included by unchecking the option “Copy the `main` branch only”.

If you think your new configuration provides significant value to the broader community, refer to the respective model configuration documentation for instructions on how to have it officially supported by ACCESS-NRI.

### Model configurations stored on MOSRS

There are two ways of getting a local copy of a configuration hosted on MOSRS:

- [Local-only copy](#rosie-checkout)
- [Local and remote copy (new remote configuration)](#rosie-copy)

Both options use [Rosie](https://metomi.github.io/rose/doc/html/tutorial/rose/furthertopics/rosie) - a tool that uses [SVN](https://subversion.apache.org) to manage, develop, and collaborate on Rose modelling workflows. It is automatically available within the [_Rose_ setup](#rose).

If you're not sure which option to use, we recommend using the "local-only copy". The "remote and local copy" is best used if you plan to commit the suite back to the remote.

> Configurations copied from MOSRS are created by default in the user's _Gadi_ home directory under `~/roses/<suite-ID>` (this is the configuration directory).

Reviewer: ```suggestion
Configurations copied from MOSRS are created by default in the user's _Gadi_ home directory under `~/roses/<suite-ID>`, which is the _configuration_ directory.
```

#### Local-only copy {: #rosie-checkout }

```
rosie checkout <suite-id>/<branch>
```

where `<suite-id>` and `<branch>` are specific to the chosen model configuration and can be found in the respective [Run a Model](/models/run_a_model/) documentation. This creates a local copy of the configuration, which is placed in the `~/roses/<suite-id>` folder.

!!! tip
    To copy from the default branch (`trunk`), omit the `/<branch>` portion of the command.

> Configurations obtained in this way cannot be pushed back to the remote. Therefore, the use of this command is recommended for testing and examining configurations.

Reviewer: ```suggestion
Configurations obtained this way cannot be pushed back to the remote repository. This command is therefore recommended for testing and inspecting configurations.
```

#### Local and remote copy (new remote configuration) {: #rosie-copy }

Before creating a new remote copy of the configuration, please read [these guidelines](https://code.metoffice.gov.uk/trac/roses-u/) on what should be stored in the Rosie repository.

```
rosie copy <suite-id>/<branch>
```

where `<suite-id>` and `<branch>` are specific to the chosen model configuration and can be found in the respective [Run a Model documentation](/models/run_a_model/).

!!! tip
    To copy from the default branch (`trunk`), omit the `/<branch>` portion of the command.

> After running this command, a text editor will open in your terminal, where you can define metadata for the new configuration (the default text editor is _Vim_, and [this quick guide](https://eastmanreference.com/a-quick-start-guide-for-beginners-to-the-vim-text-editor) is a good reference if you're unfamiliar with it). The metadata fields are expressed as `key=value` pairs, pre-filled with values copied from the original configuration. You can modify these values or add new metadata as needed. Note that `owner`, `project` and `title` are required keys. 

Reviewer: ```suggestion
After running this command, a text editor opens in your terminal, where you can define metadata for the new configuration. The default text editor is _Vim_; see [this quick guide](https://eastmanreference.com/a-quick-start-guide-for-beginners-to-the-vim-text-editor) if you are unfamiliar with it. 
Metadata fields are specified as `key=value` pairs and are pre-filled with values copied from the original configuration. You can modify these values or add new ones as needed. The `owner`, `project` and `title` keys are required. 
```

```
owner=<MOSRS-username>
project=<project-name>
title=<suite-title>
```

> When you exit the editor, you will have to confirm that you want to copy the suite:

Reviewer: ```suggestion
After exiting the editor, confirm that you want to copy the suite:
```

<terminal-window>
    <terminal-line data="input">rosie copy &lt;suite-id&gt;/&lt;branch&gt;</terminal-line>
    <terminal-line>Copy "&lt;suite-id&gt;/&lt;branch&gt;@&lt;revision&gt;" to "u-?????"? [y or n (default)]</terminal-line> <terminal-line data="input">y</terminal-line>
    <terminal-line>[INFO] &lt;new-suite-id&gt;: created at https://code.metoffice.gov.uk/svn/roses-u/&lt;suite-n/a/m/e/&gt;</terminal-line>
    <terminal-line>[INFO] &lt;new-suite-id&gt;: copied items from &lt;suite-id&gt;/&lt;branch&gt;@&lt;revision&gt;</terminal-line>
    <terminal-line>[INFO] &lt;new-suite-id&gt;: local copy created at &lt;$HOME&gt;/roses/&lt;new-suite-id&gt;</terminal-line>
</terminal-window>

> This creates a new remote configuration with a new `suite_id` (based off the copied configuration) and clones a copy of it locally in the `~/roses/<new-suite-id>` folder. Configurations created in this way are separate from the original copied configuration and can be modified and pushed back to the remote.

Reviewer: ```suggestion
This creates a new remote configuration with a new `suite_id` based on the copied configuration and clones a local copy to `~/roses/<new-suite-id>`. The new configuration is independent of the original and can be modified and pushed back to the remote.
```

To push a configuration back to the remote, from within the configuration directory run:

```
fcm commit
```

For additional `rosie` options, run:
```
rosie help
```

## Run the model configuration
!!! warning
    Before running a configuration, make sure to follow the initial setup for it (e.g., setting the correct compute project and storage resources). For details, follow the instructions relative to your specific model in the [Run a model](/models/run_a_model) page.

> ACCESS model configurations run on [_Gadi_][gadi] through [PBS jobs][PBS job] submissions. They often comprise several tasks, such as checking out code repositories, compiling and building different model components, running the model, etc. The workflow of these tasks is controlled by _Cylc_.

Reviewer: ```suggestion
ACCESS model configurations run on [_Gadi_][gadi] through [PBS jobs][PBS job] submissions. They often comprise several tasks, such as checking out code repositories, compiling and building different model components, and running the model. _Cylc_ controls the workflow and sequencing of these tasks.
```

To run the configuration, execute the following commands:

```
rose suite-run -C ~/roses/<suite-id>
```

> This launches the _Rose/Cylc_ configuration and opens the _Cylc_ GUI (if you are running on the login node and the GUI doesn't open, make sure you connected with [X11 forwarding](#x11-forwarding) enabled). The _Cylc_ GUI allows you to view and control the different tasks in the configuration as they run. The _Cylc_ GUI can be safely closed without impacting the model run. If you closed the GUI and want to re-open it, run:

Reviewer: ```suggestion
This launches the _Rose/Cylc_ configuration and opens the _Cylc_ GUI. If the GUI does not open when running on the login node, ensure you connected with [X11 forwarding](https://github.com/ACCESS-NRI/ACCESS-Hive-Docs/pull/1238/changes#x11-forwarding) enabled. The _Cylc_ GUI allows you to monitor and control tasks as they run and can be safely closed without affecting the model run. To reopen the GUI, run:
```

```
rose suite-gcontrol --name=<suite-id> &
```

!!! tip
    The `&` is optional. It detaches the invoked process, allowing the terminal prompt to remain active while the GUI is open.

By default, the configuration, log files and outputs are copied to `/scratch/<project>/${USER}/cylc-run/<suite-id>`. A symbolic link to this directory is also created in your home directory under `~/cylc-run/<suite-ID>`. See the respective [Run a model](/models/run_a_model/) documentation for details on what outputs are generated and where to find them.

## Edit the model configuration

In general, the configurations of ACCESS models can be edited either by directly modifying the configuration files within the configuration directory, or by using the [_Rose_ GUI](#rosegui).

### Rose GUI

To edit the model configuration, run the following:

```
rose edit -C ~/roses/<suite-id> &
```

> This opens the _Rose_ GUI, where the configuration settings can be modified. Once you are happy with the changes, save the modified configuration by clicking on the _Save_ button. ![Save button](/assets/run_access_cm/save_button.png){: style="height:1em"}

Reviewer: ```suggestion
This opens the _Rose_ GUI, where you can modify the configuration settings. Once you are satisfied with your changes, click the _Save_ button to save the modified configuration. 
![Save button](/assets/run_access_cm/save_button.png){: style="height:1em"}
```

!!! tip
    The `&` is optional. It detaches the invoked process, allowing the terminal prompt to remain active while the GUI is open.

!!! warning

>     Directly modifying configuration files with an editor is usually discouraged for non-expert users.

Reviewer: ```suggestion
    Directly editing configuration files is generally discouraged for non-expert users.
```
